### ISSUE TITLE
python: helloworld server is stuck if Exception is raised

### ISSUE THREAD

**Issue by attila123** [grpc/grpc#17214]
### What version of gRPC and what language are you using?
python, grpcio==1.16.0 (also tried with 1.16.1rc1)
 
### What operating system (Linux, Windows, …) and version?
recent Arch Linux
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
python 3.7.1, packages installed from pypi with pip, so gcc is not relevant I guess. Not sure which gcc version was used to compile python.
 
### What did you do?
I ran modified versions of the helloword greeter_server.py (based on https://grpc.io/docs/quickstart/python.html).

If I run it as it is (make no change to greeter_server.py) -> works -> OK.

If I run it without the loop that makes it sleep, the application exits immediately -> OK.
Code:
```
def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(), server)
    server.add_insecure_port('[::]:50051')
    server.start()
    # try:
    #     while True:
    #         time.sleep(_ONE_DAY_IN_SECONDS)
    # except KeyboardInterrupt:
    #     server.stop(0)
```

If I add a `raise Exception()` after `server.start()` then the application is stuck, does not exits -> I don't think it's OK. (I needed to Ctrl + c it.)
Code:
```
def serve():
    server = grpc.server(futures.ThreadPoolExecutor(max_workers=10))
    helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(), server)
    server.add_insecure_port('[::]:50051')
    server.start()
    raise Exception()
    # try:
    #     while True:
    #         time.sleep(_ONE_DAY_IN_SECONDS)
    # except KeyboardInterrupt:
    #     server.stop(0)
```
 
### What did you expect to see?
I would have expected that the traceback of the exception is printed then the application exists (with non zero status).
 
### What did you see instead?
The traceback of the exception is printed, but the application does not exit!
 
Based on https://github.com/grpc/grpc/blob/master/TROUBLESHOOTING.md I ran it with:
```
$ GRPC_VERBOSITY=debug GRPC_TRACE=api python greeter_server.py
I1114 10:17:36.425407004   21045 ev_epoll1_linux.cc:116]     grpc epoll fd: 3
D1114 10:17:36.425445418   21045 ev_posix.cc:169]            Using polling engine: epoll1
D1114 10:17:36.425489348   21045 dns_resolver.cc:338]        Using native dns resolver
I1114 10:17:36.425555035   21045 init.cc:153]                grpc_init(void)
I1114 10:17:36.425575457   21045 completion_queue.cc:474]    grpc_completion_queue_create_internal(completion_type=0, polling_type=0)
I1114 10:17:36.425588205   21045 init.cc:153]                grpc_init(void)
I1114 10:17:36.425601930   21045 server.cc:994]              grpc_server_create((nil), (nil))
I1114 10:17:36.425612410   21045 server.cc:979]              grpc_server_register_completion_queue(server=0x56498c9c1680, cq=0x56498c991230, reserved=(nil))
I1114 10:17:36.426177728   21045 server_chttp2.cc:34]        grpc_server_add_insecure_http2_port(server=0x56498c9c1680, addr=[::]:50051)
I1114 10:17:36.426256941   21045 socket_utils_common_posix.cc:310] TCP_USER_TIMEOUT not supported for this platform
I1114 10:17:36.426316808   21045 init.cc:153]                grpc_init(void)
I1114 10:17:36.426323379   21045 completion_queue.cc:474]    grpc_completion_queue_create_internal(completion_type=0, polling_type=1)
I1114 10:17:36.426329563   21045 server.cc:979]              grpc_server_register_completion_queue(server=0x56498c9c1680, cq=0x56498c98eb30, reserved=(nil))
I1114 10:17:36.426333922   21045 server.cc:1089]             grpc_server_start(server=0x56498c9c1680)
I1114 10:17:36.426348560   21045 completion_queue.cc:956]    grpc_completion_queue_next(cq=0x56498c98eb30, deadline=gpr_timespec { tv_sec: 1542187056, tv_nsec: 426342964, clock_type: 1 }, reserved=(nil))
I1114 10:17:36.427473583   21045 init.cc:153]                grpc_init(void)
I1114 10:17:36.427486052   21045 init.cc:153]                grpc_init(void)
I1114 10:17:36.427491224   21045 call_details.cc:31]         grpc_call_details_init(cd=0x7f7087f83270)
I1114 10:17:36.427496214   21045 metadata_array.cc:29]       grpc_metadata_array_init(array=0x7f7087fb0cd8)
I1114 10:17:36.427501604   21045 server.cc:1417]             grpc_server_request_call(server=0x56498c9c1680, call=0x7f7087f5c480, details=0x7f7087f83270, initial_metadata=0x7f7087fb0cd8, cq_bound_to_call=0x56498c991230, cq_for_notification=0x56498c991230, tag=0x7f7087fb0ca8)
I1114 10:17:36.427664089   21049 completion_queue.cc:956]    grpc_completion_queue_next(cq=0x56498c991230, deadline=gpr_timespec { tv_sec: 1542187056, tv_nsec: 627660221, clock_type: 1 }, reserved=(nil))
Traceback (most recent call last):
  File "greeter_server.py", line 47, in <module>
    serve()
  File "greeter_server.py", line 38, in serve
    raise Exception()
Exception
I1114 10:17:36.433336306   21045 server.cc:1253]             grpc_server_shutdown_and_notify(server=0x56498c9c1680, cq=0x56498c991230, tag=0x7f70886d7b48)
I1114 10:17:36.433377409   21045 completion_queue.cc:692]    cq_end_op_for_next(cq=0x56498c991230, tag=0x7f7087fb0ca8, error={"created":"@1542187056.433367027","description":"Server Shutdown","file":"src/core/lib/surface/server.cc","file_line":1291}, done=0x7f70895dccb0, done_arg=0x56498c999ff0, storage=0x56498c99a028)
I1114 10:17:36.433419365   21049 completion_queue.cc:1056]   RETURN_EVENT[0x56498c991230]: OP_COMPLETE: tag:0x7f7087fb0ca8 ERROR
I1114 10:17:36.433427873   21045 completion_queue.cc:692]    cq_end_op_for_next(cq=0x56498c991230, tag=0x7f70886d7b48, error="No Error", done=0x7f70895ded50, done_arg=0x56498c9c1680, storage=0x56498c9bd190)
I1114 10:17:36.433471110   21045 server.cc:1312]             grpc_server_cancel_all_calls(server=0x56498c9c1680)
^C
(needed to Ctrl +c it)
```

**Comment by lidizheng** [comment]
Confirm the behavior in Python 3.7.1 with gRPC 1.16.0. In Python 2 the program exit properly, I presume it might due to threading behavior difference between Python 2 and Python 3.

**Comment by lidizheng** [comment]
The gRPC Python will create 3 child threads for C-Core and one for server serving. In Python 3, the serving thread end, and main thread hung forever. There is a deadlock indeed.

Thanks to @ericgribkoff, I found the cause here is that a SHUTDOWN_TAG being added [_server.py#L757](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_server.py#L757) to completion queue, but never got propagated [_server.py#L708](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_server.py#L708) from C.

In Python 2
```
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff4686700 (LWP 189534)]
[New Thread 0x7ffff3e85700 (LWP 189535)]
[New Thread 0x7ffff3684700 (LWP 189536)]
[New Thread 0x7ffff2e83700 (LWP 189537)]
Traceback (most recent call last):
  File "greeter_server.py", line 47, in <module>
    serve()
  File "greeter_server.py", line 38, in serve
    raise Exception()
Exception
[Thread 0x7ffff3684700 (LWP 189536) exited]
[Thread 0x7ffff3e85700 (LWP 189535) exited]
[Thread 0x7ffff4686700 (LWP 189534) exited]
[Thread 0x7ffff2e83700 (LWP 189537) exited]
```

In Python 3
```
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib/x86_64-linux-gnu/libthread_db.so.1".
[New Thread 0x7ffff2505700 (LWP 190107)]
[New Thread 0x7ffff1d04700 (LWP 190108)]
[New Thread 0x7ffff1503700 (LWP 190109)]
[New Thread 0x7ffff0d02700 (LWP 190110)]
Traceback (most recent call last):
  File "greeter_server.py", line 47, in <module>
    serve()
  File "greeter_server.py", line 38, in serve
    raise Exception()
Exception
[Thread 0x7ffff0d02700 (LWP 190110) exited]
```

It might related to the object destroy behavior difference between Python 2 and Python 3. Still investigating...

**Comment by lidizheng** [comment]
Related Issue: Investigate server shutdown semantics in Cython code and possibly below #6597

**Comment by lidizheng** [comment]
Posting related info:
* One of many thread behavior difference between 2/3 https://bugs.python.org/issue31516
* [PEP 442](https://www.python.org/dev/peps/pep-0442/) -- Safe object finalization

**Comment by attila123** [comment]
@lidizheng Thanks for your investigations.
For a workaround I put my application code following the starting the grpc server to a `try` + `except Exception`, and do an `os._exit()` in the `except` body, e.g. `os._exit(1)` (I found that in https://stackoverflow.com/questions/1489669/how-to-exit-the-entire-application-from-a-python-thread).

**Comment by ericgribkoff** [comment]
Thanks for reporting this issue. The hang occurs because of the following interaction between the application's main thread and the daemon thread spawned by gRPC Python to poll the underlying C-Core server.

In the main thread:

1. When the exception is thrown, the Python interpreter begins to shutdown. 
1. This triggers our [`grpc._Server.__del__()`](https://github.com/grpc/grpc/blob/5fed8e54528ca94213fec7038f56d9db91709c33/src/python/grpcio/grpc/_server.py#L830) method to be called
1. This in turn calls sends [`_stop`](https://github.com/grpc/grpc/blob/5fed8e54528ca94213fec7038f56d9db91709c33/src/python/grpcio/grpc/_server.py#L749), which sends a shutdown request to the wrapped gRPC C-Core server, then [waits](https://github.com/grpc/grpc/blob/5fed8e54528ca94213fec7038f56d9db91709c33/src/python/grpcio/grpc/_server.py#L775) for that operation to complete asynchronously. 

This means that program exit is blocked until the core server shutdowns and the `threading.Event` is set. This is supposed to occur by means of the daemon thread started [here](https://github.com/grpc/grpc/blob/5fed8e54528ca94213fec7038f56d9db91709c33/src/python/grpcio/grpc/_server.py#L787). However, what actually happens is the following:

1. The daemon thread repeatedly [polls](https://github.com/grpc/grpc/blob/5fed8e54528ca94213fec7038f56d9db91709c33/src/python/grpcio/grpc/_server.py#L708) a completion queue, which in normal operation would soon return an event from the core server indicating that shutdown is complete
1. The call from Cython into C-Core occurs [here](https://github.com/grpc/grpc/blob/master/src/python/grpcio/grpc/_cython/_cygrpc/completion_queue.pyx.pxi#L38), with the GIL released.
1. The shutdown event pops out of the C-core completion queue as expected, leading to the exit of the `with nogil` block in our Cython code
1. Re-acquiring the GIL means invoking [PyEval_RestoreThread](https://docs.python.org/3/c-api/init.html#c.PyEval_RestoreThread)
1. ...and in particular, [this conditional](https://github.com/python/cpython/blob/989052047eea7f35da0d7ca268791b2442ee1553/Python/ceval.c#L270) which, having noticed that the interpreter is shutting down, immediately exits the thread.

As a result, the main thread hangs forever, waiting for `threading.Event.set()` from a thread that, as soon as it tries to enter back into Python code, ceases to exist. This particular CPython behavior was introduced in [this changeset](https://hg.python.org/cpython/rev/2a19d09b08f8), which explains why the example in this issue doesn't hang in Python2. However, given that the changeset fixed some serious sounding bugs with daemon threads and a shutting down interpreter, I'm not convinced that we actually *work* in this scenario even in Python2, as opposed to just getting lucky with this one particular repro case.

As with https://github.com/grpc/grpc/issues/17093, the root cause are our server deallocation methods,  blocking while waiting for a signal from threads that no longer exist. We will have a fix for this issue shortly.

**Comment by attila123** [comment]
@lidizheng @ericgribkoff thanks for your efforts.

**Comment by ericgribkoff** [comment]
The fix for this (#17444) is now merged and will be available in the next release.
